Note Widget

charliescheer opened the pull request:
### Fix
This PR is the next step in the widgets project.  Here I have added the ability to connect to the shared database with intents and widgets, parse those details, and then display them on the note widget.

<img height="400" src="https://cdn-std.droplr.net/files/acc_593859/cl7N41" />



### Test
1. Open simplenote while logged out
2. try to setup the note widget and confirm that you get an error
3. log into simplenote
4. return to the widget setup and tap on the Note option, you should see a list of all of your not deleted notes
5. choose a note with a title and some content in it
6. tap outside of the setup to return to the homescreen
7. after a second your note content should appear

### Review
***(Required)*** Add instructions for reviewers.  For example:
> Only one developer and one designer are required to review these changes, but anyone can perform the review.

### Release
> These changes do not require release notes.

charliescheer committed:
Updated title for confirmation alert

charliescheer committed:
Removed spinnerViewController ivar from settings controller

charliescheer committed:
Refactored remote error to enum instead of struct

charliescheer committed:
Updated some values to use clearer property names

charliescheer committed:
Refactored SpinnerViewController for better swift conventions

charliescheer committed:
Refactored RemoteResult random to use bool.random instead arc4

jostnes committed:
add bulk trash test

charliescheer committed:
Added defaults property that tracks if user is logged in or not

charliescheer committed:
Added StorageError enum

charliescheer committed:
Updated intents to use WidgetDataController to fetch notes at setup

charliescheer committed:
Removed WidgetState from project

charliescheer committed:
Added ability to display a note with a simperium id in widgets

charliescheer committed:
Cleaned up some classes and added notes

charliescheer committed:
Changed the name of the note preview widget to note widget

charliescheer committed:
Added demo content to display in note widget

charliescheer committed:
Updated note widget ui to design spec

charliescheer committed:
Added demo content to provider for previewing

charliescheer committed:
Fixed a UI issue where the text wouldn't reach the bottom of note widget

charliescheer committed:
Removed a few magic numbers

jostnes committed:
add dismiss modal for certain emails

charliescheer committed:
Added steps to reload widgets on timeline and when app enters background

charliescheer committed:
Handling merge conflicts with develop

charliescheer committed:
Added url linking to note widget

jleandroperez commented on SimplenoteWidgets/Tools/WidgetDataController.swift:
```diff
@@ -0,0 +1,88 @@
+        }
+
+        self.coreDataManager = coreDataManager
+        coreDataManager.managedObjectContext.persistentStoreCoordinator = coreDataManager.persistentStoreCoordinator
```
This shouldn't be done in here. This specific initialization probably deserves a new CoreDataManager initializer?

jleandroperez commented on SimplenoteWidgets/Providers/NoteWidgetProvider.swift:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        // Prepare data controller
+        var dataController: WidgetDataController
```
```suggestion
        let dataController: WidgetDataController
```

jleandroperez commented on Simplenote/Options.swift:
```diff
@@ -68,6 +68,7 @@ extension Options {
         }
         set {
             defaults.set(newValue.rawValue, forKey: .listSortMode)
+            UserDefaults(suiteName: SimplenoteConstants.sharedGroupDomain)?.set(newValue.rawValue, forKey: .listSortMode)
```
so if the Defaults aren't initialized with `suiteName`, the value isn't accessible?

Problem with this approach is: unless you set the `listSortMode` again, then, it won't be accessible by the extension.

charliescheer replied:
Good point.  I added a method to sync the setting in `applicationDidEnterBackground` as well.  That should solve the issue of the setting not syncing to the widgets/intents until it is updated

charliescheer committed:
Fixed a bug where the widget note urls were not working

charliescheer committed:
Changed accountDeletionController to be a lazy variable in app delegate

charliescheer committed:
Fixed colors for SpinnerViewController to work for dark mode

charliescheer committed:
Refactored RemoteError equality check

charliescheer committed:
Refactored RemoteError to use more specific cases

charliescheer committed:
Removed unused class declaration in settingsViewController.h

charliescheer committed:
Refactored SPAuthError to reflect additional RemoteError cases

charliescheer committed:
Removed magic number from SpinnerVC

charliescheer committed:
Refactored CoreDataManager to be able to connect the MOC and PSC at init

charliescheer committed:
Removed connection of core data manager properties from widget data

charliescheer committed:
Fixed an issue where the sort setting would only sync to widgets on set

charliescheer committed:
Added method to sync widget defaults when app becomes active/enters bg

charliescheer committed:
Updated name for accountDeletionURL constant

charliescheer committed:
Merge pull request #1354 from Automattic/issue/1353-in-app-account-deletion

In app account deletion

charliescheer committed:
Fixed colors for Note Widget to match designs

charliescheer committed:
Fixed an issue where new lines were not appearing in the note widget

charliescheer committed:
Refactored Core Data Manager init to be readable

charliescheer committed:
Refactored App Delegate methods to a widget controller class

charliescheer committed:
Removed NSObject from widget controller

charliescheer committed:
Refactored Intents creation of widget notes

charliescheer committed:
Moved core data manager to a stored property for note intent provider

charliescheer commented:
@jleandroperez I made the changes that we discussed over slack.  This is ready for another look

jostnes committed:
update review comments

jostnes committed:
use trash assert instead of note list assert

pachlava committed:
Merge pull request #1378 from jostnes/bulk-trash-test

Add bulk trash test

AliSoftware committed:
Merge pull request #1373 from Automattic/tooling/hotfix-lanes

[Tooling] Fix hotfix lanes

charliescheer committed:
Fixed a bug where the account deletion token was  created incorrectly

charliescheer committed:
Removed URL request and URL creation guards for remotes

charliescheer committed:
Removed a call to set the account deletion controller to nil

charliescheer committed:
Added ListWidgetIntents

charliescheer committed:
Added widgetList intent handler methods

charliescheer committed:
Added basic setup to supply a list widget

charliescheer committed:
Added first steps for list widget ui

SylvesterWilmott commented:
Looking good @charliescheer 

When first adding the widget, are we able to show the last modified note already in the widget? I think this experience would be better than seeing only the placeholder when first adding the widget.

I also noticed we have some space below the title. Can we remove this space? Often users ad an empty line under the note title which will result in a double empty line in the widget.

charliescheer commented:
@SylvesterWilmott yes, we can remove the padding below the title, that is easily done.

Choosing a default note... I will have to look into that.  The setting up of a widget and the widget presenting details are completely separate processes.... I don't know if there is a way to give it a default.  But I will look and see if possible

charliescheer committed:
Refactored remote url requests to be non optional

charliescheer committed:
Removed unneeded setting of the account deletion request date

charliescheer committed:
Refactored the guard checks on account deletion controller deletion bool

charliescheer committed:
Refactored intent handler to store the core data manager/data controller

charliescheer committed:
Added property to create a limited title from note

charliescheer committed:
Added default setup for note widget to auto display top of list note

charliescheer committed:
Removed header padding from note widget

charliescheer committed:
Added check in NoteProvider to confirm if timeline is for previews

charliescheer commented:
@SylvesterWilmott Turns out yes it is pretty easy to setup a default value for widgets.  It is currently setup to display the note from the top of the user's list (will use the user's sort settings to figure that out)

Also removed the padding.  This is ready to be checked again

charliescheer committed:
Changed how data controller is built for swift ui previews

charliescheer committed:
Fixed an error that  caused swift ui previews to not work

charliescheer committed:
Added url property to Note and Note widget provider

charliescheer committed:
Added logic to allow swift ui previews to build

charliescheer committed:
Added details to list widget entry

charliescheer committed:
Added demo content to use in ui views

charliescheer committed:
Added links and note content to list view

charliescheer committed:
Removed method level instance of the data managers in list intent handle

charliescheer committed:
Added default for list widget to show all notes

charliescheer committed:
Changed the name of the tag intent stored property

charliescheer committed:
Updated demo content to use a stored url

charliescheer committed:
Updated list widget to take url

charliescheer committed:
Added link in new note button view

charliescheer committed:
Added case to predicate for notes to get all notes

oguzkocer committed:
Update translations

oguzkocer committed:
Update metadata translations

oguzkocer committed:
Bump version number

oguzkocer committed:
Sanitize app store keywords after updating metadata

charliescheer committed:
Updated release notes for 4.42

oguzkocer committed:
Merge pull request #1384 from Automattic/issue/release-notes-4.42

Updated release notes for 4.42

SylvesterWilmott commented:
Thanks @charliescheer 

I think there is an issue in displaying notes with checklists that wasn't present prior to the latest commits. Notes with checklists seem to have a large space to the left of the text and checklist icons appear in plain text. Do you think we can include checklist icons here as well as eliminating the space on the left?

<img width="358" alt="Screenshot 2021-08-06 at 10 47 31" src="https://user-images.githubusercontent.com/36532468/128492158-80358e4c-8633-4eee-8a6c-591145f0d511.png">

Also, when you visit the UI where you choose from our list of widgets, it would be great to have the content from the first note in the users list populating those previews - similar to Apple Notes.

<img width="369" alt="Screenshot 2021-08-06 at 10 53 41" src="https://user-images.githubusercontent.com/36532468/128492988-113f26ea-94ab-41ea-9be6-f1c6f1ae9196.png">

jostnes committed:
wip paste not working

jostnes committed:
undo param rename

charliescheer committed:
Merge pull request #1381 from Automattic/issue/1353-account-deletion-bug

Fixed a bug where the account deletion token was  created incorrectly

jostnes committed:
add verification for internal link

charliescheer committed:
Added some modifiers to swiftUIs text struct

charliescheer committed:
Added filling modifier to view

charliescheer committed:
Added new views and modifiers to note widget

charliescheer committed:
Added ability to fetch the user's first note for note widget placeholder

charliescheer committed:
Added to the preview demo content

charliescheer commented:
@SylvesterWilmott Thanks for that feedback!

>Do you think we can include checklist icons here as well as eliminating the space on the left?

Odd that it wasn't doing this before but is now.  I have added some rules to the UI which makes the content fill the space, so that large space will not appear anymore.

As for adding the checklists, that is a bit more complicated.  Currently we generate the checklists by inserting them into AttributedStrings, but widgets can't read attributed strings by default.  We should be able to create a new SwiftUI view that will be able to process the checkboxes as they are, but that will take a bit of time to generate.  I wonder if we should make this a GH issue and handle it after the widgets are released.  WDYT?

> when you visit the UI where you choose from our list of widgets, it would be great to have the content from the first note in the users list populating those previews - similar to Apple Notes.

Totally.  That has been done, you will now see the preview of the note which is at the top of the users note list:

<img src="https://cdn-std.droplr.net/files/acc_593859/AVLmM3" />

charliescheer committed:
Fixed some errors after rebase from note widget branch

charliescheer committed:
Refactored list widget view to fix some layout issues

charliescheer committed:
Refactored new note url link to remove some repeat code

charliescheer committed:
Refactored widget header view to use note url

charliescheer committed:
Refactored new note image to be simpler/less repeat code

charliescheer committed:
Reorganized the view files in widget section of project

charliescheer committed:
Revert "Reorganized the view files in widget section of project"

This reverts commit 7889d11bd7a81b8b0d5d706b70043342b89b1713.

charliescheer committed:
Cleaned up some code for clarity

charliescheer committed:
Added dark mode support for list widget

charliescheer committed:
Changed placeholder to fetch all notes

charliescheer committed:
Fixed an issue where the list would crash if too few notes for tag

charliescheer committed:
Refactored note+widget to include a url property from its simperium key

charliescheer committed:
Refactored creating note widget entries

charliescheer committed:
Added TagsFilter to simplify setting and fetching tag predicates

charliescheer committed:
Added All Notes as an option to the list intents

SylvesterWilmott commented:
Looks great @charliescheer 

>We should be able to create a new SwiftUI view that will be able to process the checkboxes as they are, but that will take a bit of time to generate. I wonder if we should make this a GH issue and handle it after the widgets are released.

Understood. Let's create an issue and tackle that after we release widgets. I am happy with the spot we are in with the note widget so it's a :shipit: from me.

jostnes committed:
rearrange steps, update strings

jostnes committed:
use createNote instead of populateNoteList

jostnes committed:
move function to one step

oguzkocer committed:
Merge pull request #1382 from Automattic/merge/4.41-final-to-develop

Merge 4.41 final to develop

oguzkocer committed:
Bump version number

oguzkocer committed:
Update draft release notes for 4.42.

oguzkocer committed:
Release Notes: add new section for next version (4.43)

oguzkocer committed:
Update strings for localization

oguzkocer committed:
Update release toolkit to 1.4.0

oguzkocer committed:
Update metadata strings

oguzkocer committed:
Update release toolkit to 1.4

oguzkocer committed:
Merge pull request #1386 from Automattic/release/4.42

Merge 4.42 code freeze to develop

charliescheer commented:
@SylvesterWilmott  awesome!
Here is that issue https://github.com/Automattic/simplenote-ios/issues/1388
I think I know how we can do this, it is just might take some time to get it working just right

pachlava committed:
Merge pull request #1385 from jostnes/undo-trash-and-internal-link-tests

Undo via snackbar and copy internal link tests

mokagio committed:
Switch to XCUITestsHelper WIP branch

mokagio committed:
Use XCUITestsHelper method to determine if screenshots are in dark mode

charliescheer committed:
Fixed an issue where tags aren't removed from notes when deleted

charliescheer committed:
fixed a lint error I hadn't noticed before

charliescheer committed:
Added compromised password case to SPAuthError

charliescheer committed:
Added logic to present an alert for a compromised password error

charliescheer committed:
Removed a locally defined string in the SPAuthError init

charliescheer committed:
Merge pull request #1391 from Automattic/issue/1352-tag-not-removed-from-note

Fixed an issue where tags aren't removed from notes when deleted

charliescheer committed:
Added logic to logout user when deletion request completed

charliescheer committed:
Changed removal of credentials to use authenticator.reset

charliescheer committed:
Added release notes for PR 1394

oguzkocer committed:
Fix syntax errors in AppStoreStrings.pot

oguzkocer committed:
Merge pull request #1396 from Automattic/release/4.42

Fix syntax errors in AppStoreStrings.pot

charliescheer committed:
Merge pull request #1394 from Automattic/issue/password-verification

Password verification alert

charliescheer committed:
Removed app delegate call to log user out

charliescheer committed:
Added check to log out only if the correct error is thrown

charliescheer committed:
Fixed a bug that would cause the onboarding vc to load/dismiss forever

charliescheer committed:
Fixed a bug the onboarding vc was dismissed/restored after deletion

charliescheer committed:
Added URL structure to open SN to the note list displaying a certain tag

charliescheer committed:
Changed font size for note list widget

charliescheer committed:
Added link to go to tag list when note list title tapped

charliescheer committed:
Changed size for new note button on note list

charliescheer committed:
Fixed a bug where all notes could not be displayed in note list widget

charliescheer committed:
Refactored some repeat constants into a WidgetsConstants file

Commit:
Reverts App level TokenInvalid handler

Commit:
Fixes weird alignment

charliescheer committed:
fixing a git conflict with the package.resolved file

charliescheer committed:
Fixed - the list widget was not being updated when SN to background

charliescheer committed:
Refactored widget colors to utilize color studio direct from UIColor

charliescheer committed:
Fixed a group of lint errors

charliescheer committed:
Added unverified email case to spautherror

charliescheer committed:
Added alert to auth VC to handle an unverified alert case

charliescheer committed:
Fixed an issue where one of the dividers might be too large

charliescheer committed:
Fixed a bug where non http/https urls crashed when opened from markdown

mokagio committed:
Point XCUITestHelpers to version `>= 0.2.0`

mokagio committed:
Use XCUITestsHelper to query user interface style (dark mode) (#1392)

Commit:
Revert "Fixes weird alignment"

This reverts commit 7f025c26e1a0f26fe3f1bae2bc0995853c0752f8.

Commit:
Revert "Reverts App level TokenInvalid handler"

This reverts commit 0e0a7bfa5575fe682cea04da35df7c29ed9c7651.

Commit:
Wires latest Simperium

Commit:
Restores dismissAllModalsAnimated

Commit:
Adds missing Spinner dismiss

Commit:
Relocates clearRequestToken

charliescheer committed:
Merge pull request #1397 from Automattic/issue/725-webview-crash

Fixed a bug where non http/https urls crashed when opened from markdown

charliescheer committed:
Added release notes for verify at login

Commit:
Updates Simperium to Mk 1.9.0

jleandroperez commented on Simplenote/Classes/SimplenoteConstants.swift:
```diff
@@ -26,6 +26,8 @@ class SimplenoteConstants: NSObject {
     ///
     static let simplenotePublishedBaseURL = "http://simp.ly/publish/"
 
+    static let sharedGroupDomain = "group.\(Bundle.main.rootBundleIdentifier ?? "com.codality.NotationalFlow")"
```
Note that this duplicates `defaultBundleIdentifier` in `FileManager+Simplenote` (Constants struct).

Since `SimplenoteConstants` is the proper spot to share these things, mind updating FileManager to use this? 

**PLUS:** Nipticky, for consistency with the rest, mind adding a comment?

Thank yooouuuu!!!

jleandroperez commented on Simplenote/Color+Simplenote.swift:
```diff
@@ -0,0 +1,17 @@
+extension Color {
+    /// Convenience initializers to get and use Simplenote color studio colors
+    ///
+
```
Nipticky: Extra newline

jleandroperez commented on Simplenote/Controllers/WidgetController.swift:
```diff
@@ -0,0 +1,37 @@
+                return
+            }
+
+            if widgets.contains(where: { (widget) -> Bool in
```
Nipticky: No need to specify the closure return type (it's implicit in the API signature). Plus no need to wrap `(widget)` in parentheses

jleandroperez commented on Simplenote/Controllers/WidgetController.swift:
```diff
@@ -0,0 +1,37 @@
+
+                // Add check for ListWidgetIntent when created later
+
+                return false
```
Nipticky: 

Since `widget.configuration as? NoteWidgetIntent != nil` already yields a boolean, you can collapse the entire closure as follows:

```
if widgets.contains(where: { widget in
    widget.configuration as? NoteWidgetIntent != nil
})
```

jleandroperez commented on Simplenote/Controllers/WidgetController.swift:
```diff
@@ -0,0 +1,37 @@
+
+                return false
+            }) {
+                NSLog("Reloading all widgets")
```
Let's 🔥 this one!

jleandroperez commented on Simplenote/Controllers/WidgetController.swift:
```diff
@@ -0,0 +1,37 @@
+        // Set listsortmode default
+        widgetDefaults.set(UserDefaults.standard.integer(forKey: .listSortMode), forKey: .listSortMode)
+
+        // User is logged in
```
Nipticky: No need to add comments for these (since it's slightly redundant!)

jleandroperez commented on Simplenote/Controllers/WidgetController.swift:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+
+    static func syncWidgetDefaults(isLoggedIn loggedIn: Bool) {
```
```suggestion
    static func syncWidgetDefaults(authenticated: Bool) {
```

Really, really, like, really nipticky!!

jleandroperez commented on Simplenote/Date+Simplenote.swift:
```diff
@@ -0,0 +1,10 @@
+        var components = DateComponents()
+        components.hour = hours
+        return Calendar.current.date(byAdding: components, to: self)
+
```
Nipticky: Extra newline!

jleandroperez commented on Simplenote/CoreDataManager.swift:
```diff
@@ -9,27 +15,44 @@ class CoreDataManager: NSObject {
 
-        managedObjectContext = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
-        managedObjectContext.undoManager = nil
+    private func objectContext() -> NSManagedObjectContext {
```
We should prob. rename `objectContext()`, since its purpose is not clear: it's building and returning a new instance. `objectContext` would be proper for a property name.

How abooouttttt:

- `buildMainContext`
- `newMainContext`
- `createMainContext`

?

jleandroperez commented on Simplenote/CoreDataManager.swift:
```diff
@@ -9,27 +15,44 @@ class CoreDataManager: NSObject {
+        return moc
+    }
 
+    private func storeCoordinator(with model: NSManagedObjectModel, storageURL: URL) throws -> NSPersistentStoreCoordinator {
```
Same note as `objectContext()` above. How about:

- buildStoreCoordinator
- newStoreCoordinator
- createStoreCoordinator?

TY sir!!

jleandroperez commented on Simplenote/SPAppDelegate+Extensions.swift:
```diff
@@ -436,3 +437,18 @@ extension SPAppDelegate {
+
+    @objc
+    func syncWidgetDefaults() {
+        WidgetController.syncWidgetDefaults(isLoggedIn: simperium.user != nil)
```
Actually there could be a `user`, and it could have no token.

```
let authenticated = simperium.user.authenticated
```

jleandroperez commented on Simplenote/SPAppDelegate.m:
```diff
@@ -319,6 +322,8 @@ - (void)logoutAndReset:(id)sender
                 self.bSigningUserOut = NO;
 			}];
+
+            [[[NSUserDefaults alloc] initWithSuiteName:[SimplenoteConstants sharedGroupDomain]] setValue:false forKey:@"accountIsLoggedIn"];
```
This should be `syncWidgetDefaults` instead right?

jleandroperez replied:
Actually:`syncWidgetDefaults` should be invoked in three spots?

- applicationDidEnterBackground
- simperiumDidLogin **(it's in SPAppDelegate+Extensions!)**
- simperiumDidLogout **(same as above)**

jleandroperez commented on Simplenote/StorageError.swift:
```diff
@@ -0,0 +1,35 @@
+    case fetchError
+}
+
+extension StorageError {
```
Nitpciky: Most of these errors aren't really Storage Errors:

- `appConfigurationError` is thrown when the app is logged off
- `fetchError` is thrown by WidgetDataController if fetch fails

I think we could prob. get rid of this enum. Where is it actually used? (still reading!!)

charliescheer replied:
Currently this gets called when the core data manager fails to load the persistent store, if the widget data controller fails to fetch, and if the app is not logged in.

So here is my proposal for these error comments.  First if I remove the persistentStore error case from this all of these errors are specific to the widgets.  So I could change the name of the error to WidgetError and still use the configuration and fetch errors in the places they currently are.
Then in the core data manager I can throw `throw NSError(domain: Constants.errorDomain, code: .zero, userInfo: nil)` instead

I have updated the codes for that.  WDYT?

jleandroperez replied:
Swift Errors are absolutely great. Specifically, the one thing that caught my eye is  the name. It's called `StorageError`, but it may be thrown by WidgetDataController.

StorageError would maybe imply it came from the Storage* (StorageManager perhaps?)

jleandroperez commented on SimplenoteIntents/IntentHandler.swift:
```diff
@@ -2,18 +2,53 @@ import Intents
+            notes = try dataController.notes()
+        } catch {
+            completion(nil, error)
+        }
```
Question: where is the `completion` closure used? is the Error actually passed over to the user?

- If the error is presented somehow, we should prob. have our own (rather than relaying the Core Data one).
- If not, we could just drop it, and pass an empty array?

```
guard let notes = try? dataController.notes() else {
     completion([])
     return
}
```

charliescheer replied:
What appears to the user is a generic error at widget configuration.  I don't think the user is presented any of our details, but it tells the configurator there is an error and then logs the error output.  Returning an empty array would allow the configuration to load but without their notes.

🤔 let me break the widgets so I can show this

charliescheer replied:
ah, it does display:
<img height="400" src="https://cdn-std.droplr.net/files/acc_593859/OYtqFz" />

charliescheer replied:
Will add some sort of error case to be handled for this

jleandroperez commented on SimplenoteIntents/IntentHandler.swift:
```diff
@@ -2,18 +2,53 @@ import Intents
+        do {
+            notes = try dataController.notes()
+        } catch {
+            completion(nil, error)
```
Note: regardless of whether we keep this or not, this completion runs, and then execution proceeds. This would need a `return` statement

jleandroperez commented on SimplenoteWidgets/Providers/NoteWidgetProvider.swift:
```diff
@@ -0,0 +1,91 @@
+        // Prepare timeline entry for every hour for the next 6 hours
+        // Create a new set of entries at the end of the 6 entries
+        var entries: [NoteWidgetEntry] = []
+        for int in 0..<6 {
```
Let's move the range to a constant below

jleandroperez commented on SimplenoteWidgets/Tools/DemoContent.swift:
```diff
@@ -0,0 +1,9 @@
+struct DemoContent {
+    static let singleNoteTitle = "Twelve Tone Serialism"
+    static let singleNoteContent = "The twelve-tone technique is often closely related with the compositional style, serialism. Fundamentally, twelve-tone serialism is a compositional technique where all 12 notes of the chromatic scale are played with equal frequency throughout the piece without any emphasis on any one note. For this reason, twelve-tone serialism avoids being in any key. Arnold S ...
+    static let singleNoteContentAlt = "- [] Check item one \n- [] check item two\n- [] check item three"
```
Localized?

charliescheer replied:
I'm not sure if this needs to be localized.  _**Most**_ of what this does is power the SwiftUI previews. 

I was thinking that I would remove all the UI previews, but we would still need this demo content because it servers a few other purposes.  It gives the widget placeholder some sort of shape if the placeholder appears, and if for some reason there is a failure fetching the user's most recent note then this text will appear.  My feeling is that we don't really need this to be localized, WDYT?

jleandroperez replied:
TBH if there's a chance for this to be displayed, it should be localized! (Rule of thumb is: any string that may go to screen, NSLocalized!)

Localizations are extracted by the `Scripts/localize.py`. You'll prob. neeed to patch [this spot, exactly](https://github.com/Automattic/simplenote-ios/blob/develop/Scripts/localize.py#L124) as follows:

```
    genstrings_cmd = 'genstrings -q -o "%s" `find ./Simplenote ./SimplenoteShare ./SimplenoteWidgets -name "*.m" -o -name "*.swift"`'
```

(Otherwise SimplenoteWidgets won't be picked up). Same for SimplenoteShare!!

charliescheer replied:
I wonder if we should use a different text for this demo content then.  I had been using the text out of the zeplin designs just as something to put in there, but it might make sense to actually consider what this will say.

@SylvesterWilmott can you weigh in on what you would want as far as copy for placeholder text for the widgets?  There are 3 texts I can think of that may appear to users.
Title for the note widget
body for the note widget
8 note titles for the list widget

Currently all of these are using the zeplin text, which may not translate (Like I don't know how well Twelve Tone Serialism will read to users translated into another language?)  But we can continue to use these strings if you prefer

jleandroperez reviewed (commented):
Few incoming nipticky notes! Thank you sir!!

Commit:
Updates Simperium dependency

charliescheer committed:
Merge pull request #1395 from Automattic/issue/account-deletion-logout-on-succes

Log out user account after account deletion success

charliescheer committed:
Update RELEASE-NOTES.txt

Co-authored-by: Jorge Leandro Perez <jorge.perez@automattic.com>

charliescheer committed:
Added check to confirm that 403 login response is requires verification

charliescheer committed:
Added feedback messages for requesting verification at login

charliescheer committed:
Updated text for unverified error message

charliescheer committed:
Refactored SPAuthError to use where for certain cases

charliescheer committed:
Fixed some typos/alignment issues

charliescheer committed:
Merge pull request #1398 from Automattic/issue/verify-on-login

Added error case at login if user email is not verified

charliescheer committed:
Refactored FileManager to use the SimplenoteConstants for group URL

charliescheer committed:
Refactored WidgetController for readability

charliescheer committed:
Updated the method names for the core data manager

charliescheer committed:
Refactored the error for setting up the core data manager

charliescheer committed:
Added syncWidgetDefaults to login and logout

charliescheer committed:
Refactored StorageError to WidgetError

charliescheer committed:
Added return to intent handler if notes can't be fetched

charliescheer committed:
Updated use of StorageError to call WidgetError instead

charliescheer committed:
Moved timeline entry range to a static constant

charliescheer committed:
Updated release notes for 4.43

charliescheer committed:
Updated release notes to be more clear

charliescheer committed:
Merge pull request #1400 from Automattic/issue/release-notes-4.43

Updated release notes for 4.43

jleandroperez commented on Simplenote/SPAppDelegate.m:
```diff
@@ -319,6 +322,8 @@ - (void)logoutAndReset:(id)sender
                 self.bSigningUserOut = NO;
 			}];
+
+            [[[NSUserDefaults alloc] initWithSuiteName:[SimplenoteConstants sharedGroupDomain]] setValue:false forKey:@"accountIsLoggedIn"];
```
I think this line should be removed?

jleandroperez commented on SimplenoteWidgets/Models/Note+Widget.swift:
```diff
@@ -30,3 +30,52 @@ extension Note {
+            return nil
+        }
+
+        return content.components(separatedBy: .newlines)
```
You may collapse everything into:

```
   private var lines: [String]? {
         content?.components(separatedBy: .newlines)
   }
```

jleandroperez commented on SimplenoteWidgets/Models/Note+Widget.swift:
```diff
@@ -30,3 +30,52 @@ extension Note {
+        guard let lines = lines else {
+            return Constants.defaultTitle
+        }
+        return lines.first ?? String()
```
You can simplify as follows:

```
lines?.first ?? Constants.defaultTitle
```

charliescheer replied:
This is serving two roles.  It is making sure that there is content in the note and that there is a title on the first line.  If we just check that there is a first line it could be \n which would make the note appear empty in the intents.  How about removing `private var first` and changing `var title` to:

```    var title: String {
        guard let firstLine = lines?.first,
              firstLine.count > 0 else {
            return Constants.defaultTitle
        }
        return firstLine
    }
```

That serves both roles in fewer lines/properties

jleandroperez commented on SimplenoteWidgets/Providers/NoteWidgetProvider.swift:
```diff
@@ -0,0 +1,92 @@
+
+    func getSnapshot(for configuration: NoteWidgetIntent, in context: Context, completion: @escaping (NoteWidgetEntry) -> Void) {
+        guard let dataController = dataController,
+              let note = try? dataController.firstNote() else {
```
You may collapse the two first lines as follows (Unless I'm missing something!!)

```
guard let note = try? dataController?.firstNote() else {
```

jleandroperez commented on SimplenoteWidgets/Providers/NoteWidgetProvider.swift:
```diff
@@ -0,0 +1,92 @@
+                entries.append(NoteWidgetEntry(date: date, note: note))
+            }
+        }
+        let timeline = Timeline(entries: entries, policy: .atEnd)
```
Swifty Alternative!!

```
        let entries: [NoteWidgetEntry] = Constants.entryRange.compactMap { index in
            guard let date = Date().increased(byHours: index) else {
                return nil
            }

            return NoteWidgetEntry(date: date, note: note)
        }

        let timeline = Timeline(entries: entries, policy: .atEnd)
        completion(timeline)
```

jleandroperez commented on SimplenoteWidgets/Tools/WidgetDataController.swift:
```diff
@@ -0,0 +1,100 @@
+            throw WidgetError.appConfigurationError
+        }
+
+        self.coreDataManager = coreDataManager
```
There's a bit of duplication: `self.coreDataManager = coreDataManager` is implemented twice
One alternative could be:


```
 init(coreDataManager: CoreDataManager, isPreview: Bool = false) throws {
    self.coreDataManager = coreDataManager

    if isPreview {
        return
    }

    let authenticated = UserDefaults(suiteName: SimplenoteConstants.sharedGroupDomain)?.bool(forKey: .accountIsLoggedIn) ?? false

    guard authenticated else {
        throw WidgetError.appConfigurationError
        return
    }

```

charliescheer replied:
So.... The problem here is SwiftUI previews.  SwiftUI previews have no access to UserDefaults, so by adding in this check to make sure a user is logged in, it breaks the SwiftUI previews.  The first guard to see if it is a preview is meant to exit with a setup WidgetDataController without checking if the user is logged in.  I this solution above  we would still setup the WidgetDataController if the user is logged out, but would throw an error.  I'm not sure that is what we want here, but this solution works

```
    init(coreDataManager: CoreDataManager, isPreview: Bool = false) throws {
        if !isPreview {
            guard let isLoggedIn = UserDefaults(suiteName: SimplenoteConstants.sharedGroupDomain)?.bool(forKey: .accountIsLoggedIn),
                  isLoggedIn else {
                throw WidgetError.appConfigurationError
            }
        }

        self.coreDataManager = coreDataManager
    }
```

The other solution is something I have been toying with, and that is removing SwiftUI previews from the widgets before it is merged.

jleandroperez replied:
Gotcha!! thanks for changing that!!

jleandroperez commented on SimplenoteWidgets/Tools/WidgetDataController.swift:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        return SortMode(rawValue: defaults.integer(forKey: .listSortMode))
+            ?? SortMode.alphabeticallyAscending
```
I'm starting to think we could maybe use a class called... perhaps WidgetUserDefaults (similar to Options.swift).  Or not.

charliescheer replied:
I like it in it's own class.

jleandroperez commented:
I may have found a glitch @charliescheer !! Steps:

1. Fresh install Simplenote
2. Add the Note Widget
3. Tap over the Widget
4. Simplenote will launch
5. Log into your account and wait for Sync to happen
6. Go back to the springboard
7. Long press over the widget > Edit Widget
8. Choose any of your notes

As a result, the Widget will remain in "No content" / "Placeholder mode". Pressing over the widget will not open the associated Note.


https://user-images.githubusercontent.com/1195260/130161744-afe1a8d0-f408-44d7-b9ab-257e60f53e6b.mov

jleandroperez commented:
"Glitch" 2:

1. Fresh install Simplenote
2. **DO NOT** log into your account
3. Long press over the widget > Edit Widget
4. Press over the `Choose` button

As a result, you'll get the error attached in the image below. Do you think we could customize that?

![Uploading Simulator Screen Shot - iPhone 12 - 2021-08-19 at 21.23.08.png…]()

charliescheer committed:
Fixed an issue where you might not be able to fetch widget data

charliescheer committed:
Fixed an error where the intent handler was not being built correctly

charliescheer committed:
Fixed a file name mismatch from StorageError to WidgetError

charliescheer committed:
Removed un needed clearing of the widget defaults

charliescheer committed:
Refactored note widget provider for clarity and style

charliescheer committed:
Refactored check for SwiftUI preview in widget data controller

charliescheer committed:
Added WidgetDefaults class to manage options for widgets

charliescheer committed:
Refactored note+widget properties for clarity

charliescheer committed:
Merge branch 'feature/note-widget' into feature/list-Widget

charliescheer committed:
Fixed some errors left over after merging note widget into list widget

charliescheer committed:
Replaced some local file constants with constants in WidgetConstants

oguzkocer committed:
Update translations

oguzkocer committed:
Update metadata translations

oguzkocer committed:
Bump version number

oguzkocer committed:
Sanitize app store keywords after updating metadata

jleandroperez commented on Simplenote/Controllers/WidgetController.swift:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    static func syncWidgetDefaults(authenticated: Bool) {
+        let widgetDefaults = WidgetDefaults.shared
```
Beautiful!!

jleandroperez commented on SimplenoteIntents/IntentHandler.swift:
```diff
@@ -2,18 +2,62 @@ import Intents
+
+
     override func handler(for intent: INIntent) -> Any {
+        NSLog("failed to return handler")
```
We should prob drop this one (or format this a bit more, like `[IntentHandler] Failed to return handler` ?

P.s.: Does this imply the handler failed? since it's returning `self`

charliescheer replied:
I think this got left in there cause I was doing some testing on the previous glitches and wasn't sure where the issue was yet.  Will drop it

jleandroperez commented on SimplenoteWidgets/Models/Note+Widget.swift:
```diff
@@ -30,3 +30,42 @@ extension Note {
+    }
+
+    var limitedTitle: String {
+        String(title.prefix(50))
```
Nipticky: Constant!

jleandroperez commented on SimplenoteWidgets/Tools/WidgetDataController.swift:
```diff
@@ -0,0 +1,85 @@
+    /// Returns note given a simperium key
+    ///
+    func note(forSimperiumKey key: String) -> Note? {
+        notesController.predicate = predicateForNotes()
```
Let's chat over Slack, I believe we don't really need the whole ResultsController / NSFetchedResultsController machinery

jleandroperez replied:
Quick sketch, but the same machinery, without FRC, would look as follows:

```
        let request = NSFetchRequest<Note>(entityName: Note.entityName)
        request.predicate = predicate
        request.sortDescriptors = sortDescriptors
        request.fetchLimit = limit

        // This should be an ivar?
        let mainContext = coreDataManager.managedObjectContext!
        mainContext.fetch(request)
```

jleandroperez commented on SimplenoteIntents/IntentHandler.swift:
```diff
@@ -2,18 +2,62 @@ import Intents
 extension IntentHandler: NoteWidgetIntentHandling {
     func provideNoteOptionsCollection(for intent: NoteWidgetIntent, with completion: @escaping (INObjectCollection<WidgetNote>?, Error?) -> Void) {
-        var notes: [WidgetNote] = []
+        NSLog("attempting to provied note options collection")
```
Let's drop this log!

jleandroperez reviewed (commented):
Confirming that [**Bug 1** is solved!](https://github.com/Automattic/simplenote-ios/pull/1375#issuecomment-902348107)

oguzkocer committed:
Merge pull request #1404 from Automattic/release/4.42

Merge 4.42 final to develop

charliescheer committed:
Refactored a few log statements and constants

charliescheer committed:
Refactored WidgetDataController to store just the context

charliescheer committed:
Refactored Widget Data Controller to use MOC fetch instead of FRC

charliescheer committed:
Refactored some code for cleanliness

charliescheer committed:
Extracted tagExists into it's own function in SPAppDelegate

charliescheer committed:
Localized some strings that hadn't been localized

charliescheer committed:
Updating list widget with recent note widget changes

charliescheer committed:
Refactored data controller to handle updates merged from note widget

charliescheer committed:
Rearranged widget data controller to group the methods

charliescheer committed:
Updated WidgetDataController to use  tagsFilters in list widget branch

jleandroperez commented on SimplenoteWidgets/Views/NoteWidgetView.swift:
```diff
@@ -0,0 +1,43 @@
+                        .widgetHeader(widgetFamily,
+                                      color: Color(for: colorScheme, light: .gray100, dark: .white))
+                    Text(entry.content)
+                        .subheadline(color: Color(for: colorScheme, light: .gray100, dark: .white))
```
Suggestion: For reusability purposes, you would prob. want something like:

`Color+Widget.swift`

```
extension Color {

    static var bodyTextColor: Color {
        Color(for: colorScheme, light: .gray100, dark: .white)
    }
}
```

jleandroperez commented on SimplenoteWidgets/Views/NewNoteWidgetView.swift:
```diff
@@ -21,7 +21,7 @@ struct NewNoteWidgetView: View {
             .padding(Constants.overallPadding)
         }
-        .background(Color(UIColor(studioColor: .spBlue50)))
+        .background(Color(studioColor: .spBlue50))
```
(Same note as below). You would typically want something like:

```
extension Color {

   var newNoteBackgroundColor: Color {
      Color(studioColor: .spBlue50)
   }
}
```

jleandroperez commented on SimplenoteWidgets/Tools/WidgetDataController.swift:
```diff
@@ -0,0 +1,104 @@
+
+    // MARK: Fetching
+
+    private func performFetch<T: NSManagedObject>(from request: NSFetchRequest<T>) -> [T]? {
```
❤️

jleandroperez commented on SimplenoteWidgets/Tools/WidgetDataController.swift:
```diff
@@ -0,0 +1,104 @@
+import SimplenoteSearch
+import CoreData
+
+class WidgetDataController {
```
Nipticky: Since this one manages access to `Note`, perhaps a more accurate name would be `WidgetNoteController`?  (or plural?)

jleandroperez commented on SimplenoteWidgets/Providers/NoteWidgetProvider.swift:
```diff
@@ -0,0 +1,88 @@
+    let title: String
+    let content: String
+    let url: URL
+}
```
Tip: if you add the convenience initializer in an extension, you get to keep the "default" initializer(and save some lines):

```
struct NoteWidgetEntry: TimelineEntry {
    let date: Date
    let title: String
    let content: String
    let url: URL
}

extension NoteWidgetEntry {
    static let placeholder = NoteWidgetEntry(date: Date(),
                                             title: DemoContent.singleNoteTitle,
                                             content: DemoContent.singleNoteContent,
                                             url: DemoContent.demoURL)

    init(date: Date, note: Note) {
        self.init(date: date,
                  title: note.title,
                  content: note.body,
                  url: note.url)
    }
}
```

charliescheer replied:
Neat!

jleandroperez commented on SimplenoteIntents/SimplenoteIntents.entitlements:
```diff
@@ -0,0 +1,10 @@
+<dict>
+	<key>com.apple.security.application-groups</key>
+	<array>
+		<string>group.com.codality.NotationalFlow.Alpha</string>
```
I think this should be `group.$(CFBundleIdentifier)` instead.

jleandroperez replied:
cc @oguzkocer OR @mokagio OR @jkmassel  just in case!!

mokagio replied:
Nice suggestion @jleandroperez.

The result at compile-time or (run-time 🤔 ?) would be the same as this syntax, because `group.$(CFBundleIndentifier)` resolves to the bundle identifier value of the target and build configuration combination.

Having said that, it would be useful to use this syntax because it would allow us to condense `.entitlements` files that are otherwise identical. E.g.:

https://github.com/Automattic/simplenote-ios/blob/08d15d7240a7e8d23544708969451bf6f6ff4815/SimplenoteIntents/SimplenoteIntentsRelease.entitlements#L1-L10
and 

https://github.com/Automattic/simplenote-ios/blob/08d15d7240a7e8d23544708969451bf6f6ff4815/SimplenoteIntents/SimplenoteIntentsDistribution%20Internal.entitlements#L1-L10

I think my suggestion is more appropriate for a dedicated PR, though.

mokagio replied:
I opened a PR against this branch to address an entitlements issue I noticed while researching a good answer for the question above: https://github.com/Automattic/simplenote-ios/pull/1408

In the PR, I point out an issue with the Xcode UI, which I _think_ might be due to using the `CFBundleGroup` interpolation. Pending confirmation, that might be an argument _against_ going down that route and use explicit identifiers instead.

It would be a shame, though, as it's nice to have a single source of truth.

jleandroperez commented on SimplenoteIntents/SimplenoteIntentsDebug.entitlements:
```diff
@@ -0,0 +1,10 @@
+<dict>
+	<key>com.apple.security.application-groups</key>
+	<array>
+		<string>group.com.codality.NotationalFlow.Development</string>
```
Same as above, I believe this should be `group.$(CFBundleIdentifier)`

jleandroperez commented on SimplenoteIntents/SimplenoteIntentsDistribution Internal.entitlements:
```diff
@@ -0,0 +1,10 @@
+<dict>
+	<key>com.apple.security.application-groups</key>
+	<array>
+		<string>group.com.codality.NotationalFlow.Internal</string>
```
```
group.$(CFBundleIdentifier)
```

jleandroperez commented on SimplenoteIntents/SimplenoteIntentsRelease.entitlements:
```diff
@@ -0,0 +1,10 @@
+<dict>
+	<key>com.apple.security.application-groups</key>
+	<array>
+		<string>group.com.codality.NotationalFlow</string>
```
```
group.$(CFBundleIdentifier)
```

jleandroperez reviewed (approved):
✅ Code looks sharp
✅ Verified "Adding Widget after Authenticated" flow
✅ Verified "Adding Widget before Authenticated" flow
✅ Verified that the wiget's content gets refreshed
✅ Code is sharp

Few notes added. Only important ones are:

- Please verify with our Platform9 friends that the Entitlements are correct
- Please make sure the `localize.py` script is patched, to account for `SimplenoteWidget`. Otherwise Localizable Strings will **not be picked**

:shipit: when ready!!

oguzkocer committed:
Bump version number

oguzkocer committed:
Update draft release notes for 4.43.

oguzkocer committed:
Release Notes: add new section for next version (4.44)

oguzkocer committed:
Update strings for localization

oguzkocer committed:
Update release notes for 4.43

oguzkocer committed:
Update metadata strings

charliescheer committed:
Updated localized.py to fetch the strings from the widgets extension

charliescheer committed:
Added some color variables for widgets to SwiftUI Color struct

charliescheer committed:
Refactored NoteWidgetEntry to take advantage of the default init

charliescheer committed:
Fixed issue notes created on the list widget might have an all notes tag

charliescheer committed:
Refactored NoteListTable to remove the final divider

charliescheer committed:
Removed magic numbers from note list table

charliescheer committed:
Refactored widget  colors to properties of the SwiftUI Color struct

charliescheer committed:
Refactored widget tags to use a list filter kind instead of reserved key

mokagio committed:
Manually fix incorrect new lines in .pot file

We are working on a proper fix here:
https://github.com/wordpress-mobile/release-toolkit/pull/297

oguzkocer committed:
Merge pull request #1406 from Automattic/release/4.43

Merge 4.43 code freeze to develop

AliSoftware committed:
Fix mapping in metadata script & re-run it

Ran: `./fastlane/download_metadata.swift && fl sanitize_appstore_keywords` in order to update the fastlane/metadata/<lang>/*.txt files

AliSoftware committed:
Update scripts to localize app title too

AliSoftware committed:
Add app name to source metadata files

AliSoftware committed:
Update metadata strings

Ran `fastlane update_appstore_strings version:4.43` then manually fixed the issue with single-line msgid having an extra CRLF before quote

jkmassel committed:
Merge pull request #1410 from Automattic/fix-metadata

[Tooling] Fix AppStore metadata scripts

AliSoftware committed:
Merge pull request #1411 from Automattic/release/4.43

Merge metadata updates from release/4.43 to develop

charliescheer committed:
Refactored environment check to a bool stored in ProcessInfo

charliescheer committed:
Refactored simplenote links to a static function in string

charliescheer committed:
Refactored opening to a specific tag to use SPObjectManager.tagExists

charliescheer committed:
Updated some magic numbers

charliescheer committed:
Merge branch 'feature/note-widget' into feature/list-Widget

charliescheer committed:
Fixed some duplication after merging with note widget branch

charliescheer committed:
Fixed an issue where tags could not be loaded from the list widget

charliescheer committed:
Fixed an issue with the xcode project caused by a merge issue

charliescheer committed:
Refactored the widgetResultsController to be stored where it is used

charliescheer committed:
Removed check on results controller if view mode is swift ui preview

charliescheer committed:
Refactored initializer for widget tag

charliescheer committed:
Refactored internal tag url handling

charliescheer committed:
Merge branch 'develop' into feature/note-widget

charliescheer committed:
Merge pull request #1389 from Automattic/feature/list-Widget

List Widget

charliescheer committed:
Fixed an issue where date+simplenote was duplicated in target files

charliescheer committed:
Merge branch 'feature/ios-widgets' into feature/note-widget

charliescheer commented:
I made some merging mistakes while preparing this to go back into the feature branch.  I am closing this PR in favor of #1413
